File: game/atc-core/server/runtime_sovereignty.lua
-- Phase 72: Autonomous Runtime Sovereignty, Infinite Cluster Continuity & Global Runtime Finalization

local BASE = 'http://localhost:30120/api/v1/runtime-sovereignty'

-- Runtime Sovereignty
AddEventHandler('atc:sovereignty:establish', function(data, cb)
    PerformHttpRequest(BASE, function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', json.encode(data), { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sovereignty:confirm', function(id, cb)
    PerformHttpRequest(BASE .. '/' .. id .. '/confirm', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sovereignty:challenge', function(id, cb)
    PerformHttpRequest(BASE .. '/' .. id .. '/challenge', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sovereignty:revoke', function(id, cb)
    PerformHttpRequest(BASE .. '/' .. id .. '/revoke', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sovereignty:expire', function(id, cb)
    PerformHttpRequest(BASE .. '/' .. id .. '/expire', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

-- Infinite Cluster Continuity
AddEventHandler('atc:sovereignty:cluster:register', function(data, cb)
    PerformHttpRequest(BASE .. '/cluster', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', json.encode(data), { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sovereignty:cluster:degrade', function(id, cb)
    PerformHttpRequest(BASE .. '/cluster/' .. id .. '/degrade', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sovereignty:cluster:recover', function(id, cb)
    PerformHttpRequest(BASE .. '/cluster/' .. id .. '/recover', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sovereignty:cluster:fail', function(id, cb)
    PerformHttpRequest(BASE .. '/cluster/' .. id .. '/fail', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

-- Autonomous Finalization
AddEventHandler('atc:sovereignty:finalization:initiate', function(data, cb)
    PerformHttpRequest(BASE .. '/finalization', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', json.encode(data), { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sovereignty:finalization:process', function(id, cb)
    PerformHttpRequest(BASE .. '/finalization/' .. id .. '/process', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sovereignty:finalization:complete', function(id, cb)
    PerformHttpRequest(BASE .. '/finalization/' .. id .. '/complete', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sovereignty:finalization:abort', function(id, cb)
    PerformHttpRequest(BASE .. '/finalization/' .. id .. '/abort', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sovereignty:finalization:fail', function(id, cb)
    PerformHttpRequest(BASE .. '/finalization/' .. id .. '/fail', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

-- Runtime Succession
AddEventHandler('atc:sovereignty:succession:initiate', function(data, cb)
    PerformHttpRequest(BASE .. '/succession', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', json.encode(data), { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sovereignty:succession:transfer', function(id, cb)
    PerformHttpRequest(BASE .. '/succession/' .. id .. '/transfer', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sovereignty:succession:complete', function(id, cb)
    PerformHttpRequest(BASE .. '/succession/' .. id .. '/complete', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sovereignty:succession:fail', function(id, cb)
    PerformHttpRequest(BASE .. '/succession/' .. id .. '/fail', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sovereignty:succession:revert', function(id, cb)
    PerformHttpRequest(BASE .. '/succession/' .. id .. '/revert', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

-- Coordination
AddEventHandler('atc:sovereignty:coordination:upsert', function(data, cb)
    PerformHttpRequest(BASE .. '/coordination', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', json.encode(data), { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sovereignty:coordination:suspend', function(id, cb)
    PerformHttpRequest(BASE .. '/coordination/' .. id .. '/suspend', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:sovereignty:coordination:expire', function(id, cb)
    PerformHttpRequest(BASE .. '/coordination/' .. id .. '/expire', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

-- Cleanup
AddEventHandler('atc:sovereignty:cleanup', function(thresholdMs)
    PerformHttpRequest(BASE .. '/cleanup', function() end,
        'POST', json.encode({ thresholdMs = thresholdMs or 300000 }),
        { ['Content-Type'] = 'application/json' })
end)

-- Scheduled cleanup every 5 minutes
CreateThread(function()
    while true do
        Wait(300000)
        TriggerEvent('atc:sovereignty:cleanup', 300000)
    end
end)

File: apps/api/src/routes/runtime-sovereignty.ts
import type { FastifyInstance } from 'fastify'
import type { AppContext } from '../context.js'
import {
  establishSovereigntySchema,
  registerClusterSchema,
  initiateAutonomousFinalizationSchema,
  initiateSuccessionSchema,
  upsertSovereigntyCoordinationSchema,
  cleanupSovereigntySchema,
} from '@atc/operations'

export function runtimeSovereigntyRoutes(fastify: FastifyInstance, { ctx }: { ctx: AppContext }) {
  // ── Runtime Sovereignty ──────────────────────────────────────────────────────

  fastify.post('/api/v1/runtime-sovereignty', async (req, reply) => {
    if (!ctx.runtimeSovereigntyService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = establishSovereigntySchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const { sovereigntyData, ...rest } = parsed.data
    const record = await ctx.runtimeSovereigntyService.establishSovereignty({
      ...rest,
      ...(sovereigntyData !== undefined ? { sovereigntyData } : {}),
    })
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sovereignty/:id/confirm', async (req, reply) => {
    if (!ctx.runtimeSovereigntyService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeSovereigntyService.confirmSovereignty(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sovereignty/:id/challenge', async (req, reply) => {
    if (!ctx.runtimeSovereigntyService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeSovereigntyService.challengeSovereignty(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sovereignty/:id/revoke', async (req, reply) => {
    if (!ctx.runtimeSovereigntyService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeSovereigntyService.revokeSovereignty(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sovereignty/:id/expire', async (req, reply) => {
    if (!ctx.runtimeSovereigntyService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeSovereigntyService.expireSovereignty(id)
    return reply.code(200).send(record)
  })

  fastify.get('/api/v1/runtime-sovereignty/:id', async (req, reply) => {
    if (!ctx.runtimeSovereigntyService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeSovereigntyService.getSovereignty(id)
    if (!record) return reply.code(404).send({ error: 'Not found' })
    return reply.code(200).send(record)
  })

  // ── Infinite Cluster Continuity ──────────────────────────────────────────────

  fastify.post('/api/v1/runtime-sovereignty/cluster', async (req, reply) => {
    if (!ctx.infiniteClusterContinuityService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = registerClusterSchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const { clusterData, ...rest } = parsed.data
    const record = await ctx.infiniteClusterContinuityService.registerCluster({
      ...rest,
      ...(clusterData !== undefined ? { clusterData } : {}),
    })
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sovereignty/cluster/:id/degrade', async (req, reply) => {
    if (!ctx.infiniteClusterContinuityService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.infiniteClusterContinuityService.degradeCluster(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sovereignty/cluster/:id/recover', async (req, reply) => {
    if (!ctx.infiniteClusterContinuityService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.infiniteClusterContinuityService.recoverCluster(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sovereignty/cluster/:id/fail', async (req, reply) => {
    if (!ctx.infiniteClusterContinuityService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.infiniteClusterContinuityService.failCluster(id)
    return reply.code(200).send(record)
  })

  fastify.get('/api/v1/runtime-sovereignty/cluster/:clusterId', async (req, reply) => {
    if (!ctx.infiniteClusterContinuityService) return reply.code(503).send({ error: 'Service unavailable' })
    const { clusterId } = req.params as { clusterId: string }
    const record = await ctx.infiniteClusterContinuityService.getCluster(clusterId)
    if (!record) return reply.code(404).send({ error: 'Not found' })
    return reply.code(200).send(record)
  })

  // ── Autonomous Finalization ──────────────────────────────────────────────────

  fastify.post('/api/v1/runtime-sovereignty/finalization', async (req, reply) => {
    if (!ctx.autonomousFinalizationService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = initiateAutonomousFinalizationSchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const { finalizationData, ...rest } = parsed.data
    const record = await ctx.autonomousFinalizationService.initiateFinalization({
      ...rest,
      ...(finalizationData !== undefined ? { finalizationData } : {}),
    })
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sovereignty/finalization/:id/process', async (req, reply) => {
    if (!ctx.autonomousFinalizationService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.autonomousFinalizationService.processFinalization(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sovereignty/finalization/:id/complete', async (req, reply) => {
    if (!ctx.autonomousFinalizationService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.autonomousFinalizationService.completeFinalization(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sovereignty/finalization/:id/abort', async (req, reply) => {
    if (!ctx.autonomousFinalizationService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.autonomousFinalizationService.abortFinalization(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sovereignty/finalization/:id/fail', async (req, reply) => {
    if (!ctx.autonomousFinalizationService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.autonomousFinalizationService.failFinalization(id)
    return reply.code(200).send(record)
  })

  fastify.get('/api/v1/runtime-sovereignty/finalization/:id', async (req, reply) => {
    if (!ctx.autonomousFinalizationService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.autonomousFinalizationService.getFinalization(id)
    if (!record) return reply.code(404).send({ error: 'Not found' })
    return reply.code(200).send(record)
  })

  // ── Runtime Succession ───────────────────────────────────────────────────────

  fastify.post('/api/v1/runtime-sovereignty/succession', async (req, reply) => {
    if (!ctx.runtimeSuccessionService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = initiateSuccessionSchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const { successionData, targetServerId, ...rest } = parsed.data
    const record = await ctx.runtimeSuccessionService.initiateSuccession({
      ...rest,
      ...(targetServerId !== undefined ? { targetServerId } : {}),
      ...(successionData !== undefined ? { successionData } : {}),
    })
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sovereignty/succession/:id/transfer', async (req, reply) => {
    if (!ctx.runtimeSuccessionService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeSuccessionService.beginTransfer(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sovereignty/succession/:id/complete', async (req, reply) => {
    if (!ctx.runtimeSuccessionService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeSuccessionService.completeSuccession(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sovereignty/succession/:id/fail', async (req, reply) => {
    if (!ctx.runtimeSuccessionService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeSuccessionService.failSuccession(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sovereignty/succession/:id/revert', async (req, reply) => {
    if (!ctx.runtimeSuccessionService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeSuccessionService.revertSuccession(id)
    return reply.code(200).send(record)
  })

  fastify.get('/api/v1/runtime-sovereignty/succession/:id', async (req, reply) => {
    if (!ctx.runtimeSuccessionService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.runtimeSuccessionService.getSuccession(id)
    if (!record) return reply.code(404).send({ error: 'Not found' })
    return reply.code(200).send(record)
  })

  // ── Distributed Sovereignty Coordination ────────────────────────────────────

  fastify.post('/api/v1/runtime-sovereignty/coordination', async (req, reply) => {
    if (!ctx.distributedSovereigntyCoordinator) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = upsertSovereigntyCoordinationSchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const { coordinationData, ...rest } = parsed.data
    const record = await ctx.distributedSovereigntyCoordinator.upsertCoordination({
      ...rest,
      ...(coordinationData !== undefined ? { coordinationData } : {}),
    })
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sovereignty/coordination/:id/suspend', async (req, reply) => {
    if (!ctx.distributedSovereigntyCoordinator) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.distributedSovereigntyCoordinator.suspendCoordination(id)
    return reply.code(200).send(record)
  })

  fastify.post('/api/v1/runtime-sovereignty/coordination/:id/expire', async (req, reply) => {
    if (!ctx.distributedSovereigntyCoordinator) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const record = await ctx.distributedSovereigntyCoordinator.expireCoordination(id)
    return reply.code(200).send(record)
  })

  fastify.get('/api/v1/runtime-sovereignty/coordination/:coordinationId', async (req, reply) => {
    if (!ctx.distributedSovereigntyCoordinator) return reply.code(503).send({ error: 'Service unavailable' })
    const { coordinationId } = req.params as { coordinationId: string }
    const record = await ctx.distributedSovereigntyCoordinator.getCoordination(coordinationId)
    if (!record) return reply.code(404).send({ error: 'Not found' })
    return reply.code(200).send(record)
  })

  // ── Cleanup ──────────────────────────────────────────────────────────────────

  fastify.post('/api/v1/runtime-sovereignty/cleanup', async (req, reply) => {
    if (!ctx.sovereigntyRecoveryService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = cleanupSovereigntySchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const result = await ctx.sovereigntyRecoveryService.cleanupStale(parsed.data.thresholdMs)
    return reply.code(200).send(result)
  })
}
